test/e2e/storage/csi_volumes.go

```diff
@@ -196,57 +195,52 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 	})
 
 	// The CSIDriverRegistry feature gate is needed for this test in Kubernetes 1.12.
-	Context("CSI attach test using HostPath driver [Feature:CSIDriverRegistry]", func() {
+
+	Context("CSI attach test using mock driver [Feature:CSIDriverRegistry]", func() {
 		var (
 			driver testsuites.TestDriver
 		)
 
-		BeforeEach(func() {
-			config := testsuites.TestConfig{
-				Framework: f,
-				Prefix:    "csi-attach",
-			}
-			driver = drivers.InitHostPathCSIDriver(config)
-			driver.CreateDriver()
-		})
-
 		AfterEach(func() {
 			driver.CleanupDriver()
 		})
 
 		tests := []struct {
-			name                   string
-			driverAttachable       bool
-			driverExists           bool
-			expectVolumeAttachment bool
+			name             string
+			driverAttachable bool
+			driverExists     bool
 		}{
 			{
-				name:                   "non-attachable volume does not need VolumeAttachment",
-				driverAttachable:       false,
-				driverExists:           true,
-				expectVolumeAttachment: false,
+				name:             "should not require VolumeAttach for drivers without attachment",
+				driverAttachable: false,
+				driverExists:     true,
 			},
 			{
-				name:                   "attachable volume needs VolumeAttachment",
-				driverAttachable:       true,
-				driverExists:           true,
-				expectVolumeAttachment: true,
+				name:             "should require VolumeAttach for drivers with attachment",
+				driverAttachable: true,
+				driverExists:     true,
 			},
 			{
-				name:                   "volume with no CSI driver needs VolumeAttachment",
-				driverExists:           false,
-				expectVolumeAttachment: true,
+				name:             "should preserve attachment policy when no CSIDriver present",
+				driverAttachable: true,
+				driverExists:     false,
 			},
 		}
 
 		for _, t := range tests {
 			test := t
 			It(test.name, func() {
+				By("Deploying mock CSI driver")
+				config := testsuites.TestConfig{
+					Framework: f,
+					Prefix:    "csi-attach",
+				}
+
+				driver = drivers.InitMockCSIDriver(config, test.driverExists, test.driverAttachable, nil)
+				driver.CreateDriver()
+
 				if test.driverExists {
-					csiDriver := createCSIDriver(csics, testsuites.GetUniqueDriverName(driver), test.driverAttachable, nil)
-					if csiDriver != nil {
-						defer csics.CsiV1alpha1().CSIDrivers().Delete(csiDriver.Name, nil)
-					}
+					defer destroyCSIDriver(csics, driver)
```

Comment:
Can we add `driver.CleanupDriver()` here instead of in the `AfterEach` to be consistent with the create?

Reply:
Can do.